cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -135,41 +133,80 @@ struct copy_if_else_functor {
       auto lhs_dv = column_device_view::create(lhs);
       auto rhs_dv = column_device_view::create(rhs);
       auto out_dv = mutable_column_device_view::create(*out);
-      
+
       // if we have validity in the output
       if(validity_policy == experimental::mask_allocation_policy::ALWAYS){
          rmm::device_scalar<cudf::size_type> valid_count{0, stream, mr};
-         
+
          // call the kernel
-         copy_if_else_kernel<block_size, T, Filter, true><<<grid.num_blocks, block_size, 0, stream>>>(
+         copy_if_else_kernel<block_size, Element, FilterFn, true><<<grid.num_blocks, block_size, 0, stream>>>(
             *lhs_dv, *rhs_dv, filter, *out_dv, valid_count.data());
 
          out->set_null_count(lhs.size() - valid_count.value());
       } else {
          // call the kernel
-         copy_if_else_kernel<block_size, T, Filter, false><<<grid.num_blocks, block_size, 0, stream>>>(
+         copy_if_else_kernel<block_size, Element, FilterFn, false><<<grid.num_blocks, block_size, 0, stream>>>(
             *lhs_dv, *rhs_dv, filter, *out_dv, nullptr);
       }
 
       return out;
    }
 };
 
+/**
+ * @brief Specialization functor for strings column to perform a copy if/else
+ *        using a filter function to select from lhs/rhs columns.
+ */
+template<typename FilterType>
+struct copy_if_else_functor_impl<string_view, FilterType>
+{
+  std::unique_ptr<column> operator()(column_view const& lhs,
+                                     column_view const& rhs,
+                                     FilterType filter,
+                                     rmm::mr::device_memory_resource *mr,
+                                     cudaStream_t stream)
+   {
+      return strings::detail::copy_if_else( strings_column_view(lhs),
+                                            strings_column_view(rhs),
+                                            filter, mr, stream);
+   }
+};
+
+/**
+ * @brief Functor called by the `type_dispatcher` in order to perform a copy if/else
+ *        using a filter function to select from lhs/rhs columns.
+ *
+ * This is required to split the specialization of the type Element from the Filter function.
```

Comment:
I don't think you need two layers of structs. I think you can template the struct on the filter, and use SFINAE on the operator() to specialize it for strings vs. fixed-width types.

Reply:
I tried alot of things to get this to work. I ended up copying the logic implemented here for the same purpose:
https://github.com/rapidsai/cudf/blob/f5687b2c74381e626c0021a0fd3f89ca0f40190d/cpp/include/cudf/detail/gather.cuh#L161
I can try again using `is_fixed_width` to see what error I get.

Reply:
It's not a huge deal.